### PR TITLE
Move samples to non-root images and multi-stage Dockerfiles

### PR DESCRIPTION
**Pull request by paurosello**
## Description
Improving docker images to use a non-root user and standardising the fwatchdog installation using multistage Dockerfiles

## Motivation and Context
- [x] My issue has received approval from the maintainers or lead with the `design/approved` label
https://github.com/openfaas/faas/issues/1167

## How Has This Been Tested?
- Build the image
- Start fwatchdog process
- HTTP request to 8080 port

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: sample-functions/DockerHubStats/Dockerfile [hunk 1]
```diff
@@ -1,26 +1,26 @@
...
-    && chmod +x /usr/bin/fwatchdog
-
 WORKDIR /go/src/github.com/openfaas/faas/sample-functions/DockerHubStats
-
 COPY . /go/src/github.com/openfaas/faas/sample-functions/DockerHubStats
-
-RUN make install
+RUN set -ex && apk add make && make install
 
 FROM alpine:3.8
```
Comment (alexellis) [inline sample-functions/DockerHubStats/Dockerfile]: I think `3.10` is out now?

File: sample-functions/AlpineFunction/Dockerfile [hunk 2]
```diff
@@ -8,5 +8,8 @@ RUN chmod +x /usr/bin/fwatchdog
 # Populate example here
 # ENV fprocess="wc -l"
 
+RUN addgroup -g 1000 -S faas && adduser -u 1000 -S faas -G faas
```
Comment (alexellis) [inline sample-functions/AlpineFunction/Dockerfile]: Please can you use the style from the other functions where we use the username `app`?

This is for consistency between [all templates and samples](https://github.com/openfaas/templates/blob/master/template/go/Dockerfile).

File: sample-functions/DockerHubStats/Dockerfile (hunk 1 above)
Comment (paurosello) [reply sample-functions/DockerHubStats/Dockerfile]: Moved everything to alpine:3.10

**Comment by alexellis** [comment]
Hi Pau, thanks for pushing those two changes. 

Unfortunately there are some files changed in vendor.. did you do a global search and replace? I would recommend squashing this and resetting the vendor folder at the same time. 

Alex

File: sample-functions/AlpineFunction/Dockerfile (hunk 2 above)
Comment (paurosello) [reply sample-functions/AlpineFunction/Dockerfile]: I changed the user to be app, but I leave the USER <userid> to make it compliant with [OpenShift](https://docs.openshift.com/container-platform/3.11/creating_images/guidelines.html#openshift-specific-guidelines)

We might have to modify templates too

**Comment by paurosello** [comment]
Oh didn´t see that one :) will take fix now

**Comment by alexellis** [comment]
Can we do the OpenShift suggestion separately please? If you could raise an issue so it can be managed, that would be great.

**Comment by paurosello** [comment]
Sure, will open Issue and will squash all commits once everything is fine and ready to merge

**Comment by paurosello** [comment]
@alexellis do you think this is ready?

**Comment by alexellis** [comment]
Thank you it looks great.

I'll merge, but I'd like to see `USER 1000` being `USER app`, since that is the user we added and use in all the templates.

**Comment by alexellis** [comment]
Also you didn't sign-off the commit.

Derek explains how to do that with:

```
git commit -s --amend
```

**Pull request merged**